README.md
# MEMDEM

![build](https://github.com/underchemist/memdem/actions/workflows/ci.yaml/badge.svg)
[![codecov](https://codecov.io/gh/underchemist/memdem/branch/main/graph/badge.svg?token=U2WUJUDDUS)](https://codecov.io/gh/underchemist/memdem)
[![PyPI version](https://badge.fury.io/py/memdem.svg)](https://badge.fury.io/py/memdem)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

Make use of open global elevation data from [Terrain Tiles](https://registry.opendata.aws/terrain-tiles) to create digital elevation models (DEM) on the fly.

The memdem project credits the [data source providers](ATTRIBUTION.md). Any use of memdem for redistribution, resale, presentation or publication of the data available from the Terrain Tiles dataset is required to make a similar attribution. See [here](https://github.com/tilezen/joerd/blob/master/docs/attribution.md) for more information.

# Installation

memdem depends on 

- rasterio[s3] 
- morecantile

## Linux / OSX

```
pip install memdem
```

## Windows
**note**: Rasterio does not provide official windows wheels, ensure [GDAL](https://github.com/osgeo/gdal) is installed in order to install rasterio dependency.

### Unofficial wheels

[Unofficial rasterio wheels](https://www.lfd.uci.edu/~gohlke/pythonlibs/) (includes GDAL).

### Conda

```
conda create -yn memdem python=3.9
conda activate memdem
conda install -c conda-forge gdal
```

Once GDAL is installed, the environment variable `GDAL_VERSION` may need to be set in order to install rasterio dependency.

``` 
pip install memdem
```

# Quickstart

```python
import rasterio
from rasterio.crs. import CRS

import memdem

# define a bounding box
bounds = (-123.61735447008346, 48.91185687803928, -122.74734074899379, 49.52797572603976)

# create an instance of DEMTiles
demtiles = memdem.DEMTiles(
    bounds,
    zoom=10
)

# behold, a DEM created on the fly
with demtiles.open() as dem:
    lon, lat = (-123.17911189115098, 49.26435101976154)
    point_xy = rasterio.warp.transform(CRS.from_epsg(4326), dem.crs, [lon], [lat])
    print(next(dem.sample(zip(*point_xy))))  # 20


# save DEM to file for later use
demtiles.to_file("path/to/dem.vrt")
```

# Notes

## Coordinate Reference System

Terrain Tiles are referenced to the EPSG:3857 aka Web Mercator projection. However the vertical 
datum (what reference is used for elevation values e.g. WGS84 ellipsoid) [may vary depending on the source data](https://github.com/tilezen/joerd/issues/130). For accurate elevation results, it may be required to use source data with a well-defined vertical datum. For example, SRTM data has vertical datum EGM96. 

## AWS credentials

The Terrain Tiles dataset is avaiable freely and does not require aws credentials to access. Rasterio uses signed requests by default and may result in an error if it cannot locate your aws credentials. By default, memdem wraps the the `DEMTiles.open` context in a `rasterio.Env(aws_unsigned=True)` context. For operations requiring dataset reads outside of the `DEMTiles.open` context (e.g. when opening a previously saved VRT) ensure you do so within a `rasterio.Env(aws_unsigned=True)` block.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0.post1] - 2021-07-11

## Changed
- packaging metadata to be more pypi friendly

## [0.1.0] - 2021-07-11

### Added
- Initial release.

tests/conftest.py
import rasterio
from affine import Affine

from memdem.constants import ELEVATION_TILE_SIZE
from memdem.tms import default_tms

SEED_XY = (-123.17911189115098, 49.26435101976154)
SEED_ZOOM = 10
METERS_PER_TILE_AT_Z10 = 39135.8
INT16_NODATA = -32768


def fake_mem_dataset(index=0):
    crs = rasterio.crs.CRS.from_epsg(3857)
    tms = default_tms.get("ElevationWebMercatorQuad")
    seed_tile = tms.tile(*SEED_XY, zoom=SEED_ZOOM)
    transform = rasterio.transform.from_bounds(
        *tms.xy_bounds(seed_tile), ELEVATION_TILE_SIZE, ELEVATION_TILE_SIZE
    )
    transform = transform * Affine.translation(0, METERS_PER_TILE_AT_Z10 * index)
    mem = rasterio.MemoryFile()
    src = mem.open(
        driver="GTiff",
        width=ELEVATION_TILE_SIZE,
        height=ELEVATION_TILE_SIZE,
        count=1,
        crs=crs,
        dtype="int16",
        transform=transform,
        nodata=INT16_NODATA,
    )
    return (mem, src)


tests/test_tms.py
"""Test custom TileMatrixSet."""
from rasterio.crs import CRS

from memdem.constants import ELEVATION_TILE_SIZE
from memdem.tms import default_tms


def test_elevationwebmercatorquad():
    assert "ElevationWebMercatorQuad" in default_tms.list()
    tms = default_tms.get("ElevationWebMercatorQuad")
    assert tms.crs == CRS.from_epsg(3857)
    matrix = tms.matrix(0)
    assert matrix.tileHeight == ELEVATION_TILE_SIZE
    assert matrix.tileWidth == ELEVATION_TILE_SIZE


tests/test_vrt.py
import rasterio
from affine import Affine

import memdem

from .conftest import fake_mem_dataset


def test_buildvrt():
    demtiles = memdem.DEMTiles((0, 0, 1, 1), zoom=10)
    urls = demtiles.get_tile_urls()
    vrt = memdem.vrt.buildvrt(
        urls, demtiles.shape, demtiles.offsets, demtiles.transform, demtiles.tms.crs
    )
    assert vrt.startswith("<VRTDataset")

    with rasterio.open(vrt) as src:
        assert src.count == 1
        assert src.shape == demtiles.shape
        assert src.crs == rasterio.crs.CRS.from_epsg(3857)
        assert src.transform == demtiles.transform


def test_buildvrt_axis_swap_crs():
    crs = rasterio.crs.CRS.from_epsg(4326)
    demtiles = memdem.DEMTiles((0, 0, 1, 1), zoom=10)
    urls = demtiles.get_tile_urls()
    vrt = memdem.vrt.buildvrt(
        urls, demtiles.shape, demtiles.offsets, demtiles.transform, crs
    )
    assert 'dataAxisToSRSAxisMapping="2,1"' in vrt


memdem/memdem.py
"""Module for generating DEM."""
from contextlib import contextmanager
from functools import wraps
from os import PathLike
from typing import Iterator, Optional, Sequence, Tuple, Union

import morecantile
import rasterio
import rasterio.shutil

from memdem.constants import ELEVATION_TILE_SIZE, SERVICE_S3_URL
from memdem.tms import default_tms
from memdem.vrt import buildvrt


class DEMTiles:
    """
    DEM tiles.

    Attributes
    ----------
    tms : morecantile.TileMatrixSet
        Custom TileMatrixSet for Elevation Tiles dataset. Adapted from
        WebMercatorQuad with 512x512 tile size.
    """

    tms: morecantile.TileMatrixSet = default_tms.get("ElevationWebMercatorQuad")

    def __init__(
        self,
        bounds: Union[Tuple[float, float, float, float], rasterio.coords.BoundingBox],
        zoom: int = 8,
        pixel_size: Optional[float] = None,
    ):
        """
        Define the parameters needed to build a virtual DEM.

        Parameters
        ----------
        bounds : Tuple[float, float, float, float] or rasterio.coords.BoundingBox
            The bounding box describing the area of interest in geographic coordinates.
            Tiles overlapping this bounding box will be collected for the specified
            zoom level.
        zoom : int
            Zoom level to use.
        pixel_size : float, optional
            Desired pixel size. Used to compute a suitable zoom level from
            TileMatrixSet.zoom_for_res.

        Attributes
        ----------
        tileset : Sequence of morecantile.Tile
            The tiles overlapping the bounds.
        zoom : int
            The zoom level to use, either as specified or as computed by
            TileMatrixSet.zoom_for_res using the zoom_level_strategy="upper"
        nx, ny : int
            The number of tiles in the x and y dimensions, respectively.
        """
        self.zoom = zoom
        if pixel_size:
            self.zoom = self.tms.zoom_for_res(pixel_size, zoom_level_strategy="upper")

        tileset = self.tms.tiles(*bounds, zooms=self.zoom)
        self.tileset, self.nx, self.ny = self._extract_tileset(tileset)

        self._width = self._get_width()
        self._height = self._get_height()
        self._shape = self._get_shape()
        self._offsets = self._get_offsets()
        self._transform = self._get_transform()
        self._bounds = self._get_bounds()

    @property
    def width(self):
        """Return the width of the DEM in pixels."""
        return self._width

    def _get_width(self):
        return self.nx * ELEVATION_TILE_SIZE

    @property
    def height(self):
        """Return the height of the DEM in pixels."""
        return self._height

    def _get_height(self):
        return self.ny * ELEVATION_TILE_SIZE

    @property
    def shape(self):
        """Return the (height, width) of the DEM in pixels."""
        return self._shape

    def _get_shape(self):
        return (self.height, self.width)

    @property
    def offsets(self):
        """Return a sequence of (xoffset, yoffset, xsize, ysize) corresponding to each Tile in tileset. Used to mosaic individual tiles into a single DEM."""
        return self._offsets

    def _get_offsets(self):
        offsets = []
        first_tile = next(iter(self.tileset))
        startx = first_tile.x
        starty = first_tile.y
        for tile in self.tileset:
            dst_idxx = tile.x - startx
            dst_idxy = tile.y - starty
            offsets.append(
                (
                    ELEVATION_TILE_SIZE * dst_idxx,
                    ELEVATION_TILE_SIZE * dst_idxy,
                    ELEVATION_TILE_SIZE,
                    ELEVATION_TILE_SIZE,
                )
            )
        return offsets

    @property
    def bounds(self):
        """Return bounds of tileset in tile crs."""
        return self._bounds

    def _get_bounds(self):
        left, bottom, right, top = (0, self.height, self.width, 0)
        left, bottom = self.transform * (left, bottom)
        right, top = self.transform * (right, top)
        return rasterio.coords.BoundingBox(
            left=left, bottom=bottom, right=right, top=top
        )

    @property
    def transform(self):
        """Return geotransform for tileset in tile crs."""
        return self._transform

    def _get_transform(self):
        matrix = self.tms.matrix(self.zoom)
        pixel_size = self.tms._resolution(matrix)

        ul_tile = min(self.tileset)
        coords = self.tms.ul(ul_tile)
        coords_xy = self.tms.xy(*coords)
        return rasterio.transform.from_origin(*coords_xy, pixel_size, pixel_size)

    def _extract_tileset(
        self, tiles: Iterator[morecantile.Tile]
    ) -> Tuple[Sequence[morecantile.Tile], int, int]:
        tileset = []
        x = []
        y = []
        for tile in tiles:
            tileset.append(tile)
            x.append(tile.x)
            y.append(tile.y)

        nx = len(set(x))
        ny = len(set(y))

        return (tileset, nx, ny)

    def get_tile_urls(self) -> Sequence[str]:
        """Build s3 urls from tileset."""
        return [SERVICE_S3_URL.format(z=t.z, x=t.x, y=t.y) for t in self.tileset]

    def to_string(self):
        """Generate VRT string."""
        urls = self.get_tile_urls()
        return buildvrt(urls, self.shape, self.offsets, self.transform, self.tms.crs)

    def to_file(self, fname: Union[str, PathLike]):
        """Write VRT to file."""
        rasterio.shutil.copy(self.to_string(), fname, driver="VRT")

    @contextmanager
    def open(self):
        """Yield rasterio dataset of DEM VRT."""
        with rasterio.Env(aws_unsigned=True):
            with rasterio.MemoryFile(ext=".vrt") as mem:
                rasterio.shutil.copy(self.to_string(), mem.name, driver="VRT")
                with mem.open() as dem:
                    yield dem

    @classmethod
    def from_dataset(
        cls,
        dataset: rasterio.DatasetReader,
        zoom: int = 8,
        pixel_size: Optional[float] = None,
    ):
        """Create DEMTiles using bounds defined by dataset.

        Bounds in non-geographic coordinates will be transformed automatically.

        Parameters
        ----------
        dataset : rasterio.DatasetReader
            A rasterio dataset opened in 'r' mode. Must have a geotransform
            and crs.
        zoom : int
            Zoom level to use.
        pixel_size : float, optional
            Desired pixel size. Used to compute a suitable zoom level from
            TileMatrixSet.zoom_for_res.
        """
        if dataset.transform.is_identity:
            raise ValueError("Dataset must have a geotransform")
        if not dataset.crs:
            raise ValueError("Dataset has no CRS")

        # ensure bounds in geographic (lon, lat) coordinates
        bounds = dataset.bounds
        bounds = rasterio.warp.transform_bounds(
            dataset.crs, rasterio.crs.CRS.from_epsg(4326), *bounds
        )

        return cls(bounds, zoom=zoom, pixel_size=pixel_size)


memdem/vrt.py
"""Module for VRT creation."""
from typing import Any, Iterable, Sequence, Tuple
from xml.etree import ElementTree as ET

import morecantile
import rasterio
from affine import Affine

from memdem.constants import ELEVATION_TILE_SIZE, ELEVATION_BLOCK_SIZE


def buildvrt(
    urls: Sequence[str],
    shape: Tuple[int, int],
    offsets: Sequence[Tuple[int, int, int, int]],
    transform: Affine,
    crs: rasterio.crs.CRS = rasterio.crs.CRS.from_epsg(3857),
    dtype: str = "int16",
    nodata: Any = None,
) -> str:
    """
    Create a VRT document of mosaic'd Elevation Tiles for a given zoom level,
    representing a single DEM over the area defined.

    Parameters
    ----------
    urls : Sequence of str
        Sequence of s3 urls pointing to valid Elevation Tiles, which collectively
        represent a DEM.
    shape : Two-tuple of int
        The pixel height and width of DEM.
    offsets : Sequence of Four-tuple
        Sequence of tuples of the form (xoff, yoff, xsize, ysize) mapping the individual
        tile coordinates to the DEM image coordinate space.
    transform : Affine
        The geotransform for the DEM.
    crs : rasterio.crs.CRS
        The crs of the DEM, inheriting from the crs of the TileMatrixSet.
    dtype : str, optional
        Valid string identifier of a numpy datatype, determines the DEM dtype.
        Defaults to "int16".
    nodata : int or float, optional
        The nodata value to set for the DEM. Defaults to -32768

    Returns
    -------
    str:
        An XML formatted string of a valid VRT dataset.
    """
    if nodata is None:
        nodata = min(rasterio.dtypes.dtype_ranges[dtype])
    height, width = shape
    parsed_paths = [rasterio.parse_path(ds) for ds in urls]
    vsi_paths = [p.as_vsi() for p in parsed_paths]

    # root element
    root = ET.Element("VRTDataset")
    root.set("rasterXSize", str(width))
    root.set("rasterYSize", str(height))

    # crs/SRS
    wkt = crs.to_wkt()
    axis_mapping = "1,2"
    if rasterio.crs.epsg_treats_as_latlong(
        crs
    ) or rasterio.crs.epsg_treats_as_northingeasting(crs):
        axis_mapping = "2,1"
    srs = ET.SubElement(root, "SRS")
    srs.set("dataAxisToSRSAxisMapping", axis_mapping)
    srs.text = wkt

    # geotransform
    transform_str = ", ".join(
        list(map(lambda x: "{:.16e}".format(x), transform.to_gdal()))
    )
    geotransform = ET.SubElement(root, "GeoTransform")
    geotransform.text = transform_str

    # bands
    vrtrasterband = ET.SubElement(root, "VRTRasterBand")
    vrtrasterband.set("dataType", rasterio.dtypes._gdal_typename(dtype))
    vrtrasterband.set("band", "1")
    vrtnodata = ET.SubElement(vrtrasterband, "NoDataValue")
    vrtnodata.text = str(nodata)
    vrtcolorinterp = ET.SubElement(vrtrasterband, "ColorInterp")
    vrtcolorinterp.text = "Gray"

    for offset, uri in zip(offsets, vsi_paths):
        source = ET.SubElement(vrtrasterband, "ComplexSource")
        sourcefilename = ET.SubElement(source, "SourceFilename")
        sourcefilename.set("relativeToVRT", "0")
        sourcefilename.text = uri
        sourceband = ET.SubElement(source, "SourceBand")
        sourceband.text = "1"
        sourceproperties = ET.SubElement(source, "SourceProperties")
        sourceproperties.set("RasterXSize", str(ELEVATION_TILE_SIZE))
        sourceproperties.set("RasterYSize", str(ELEVATION_TILE_SIZE))
        sourceproperties.set("DataType", "Int16")
        sourceproperties.set("BlockXSize", str(ELEVATION_BLOCK_SIZE))
        sourceproperties.set("BlockYSize", str(ELEVATION_BLOCK_SIZE))
        srcrect = ET.SubElement(source, "SrcRect")
        srcrect.set("xOff", "0")
        srcrect.set("yoff", "0")
        srcrect.set("xSize", str(ELEVATION_TILE_SIZE))
        srcrect.set("ySize", str(ELEVATION_TILE_SIZE))
        xoff, yoff, xsize, ysize = offset
        dstrect = ET.SubElement(source, "DstRect")
        dstrect.set("xOff", str(xoff))
        dstrect.set("yOff", str(yoff))
        dstrect.set("xSize", str(xsize))
        dstrect.set("ySize", str(ysize))
        srcnodata = ET.SubElement(source, "NODATA")
        srcnodata.text = str(nodata)

    return ET.tostring(root, "utf-8").decode("utf-8")
